database/database.go

```diff
@@ -193,41 +187,44 @@ func (t *Txn) Count() (count uint64, err error) {
 	return
 }
 
-// ChannelPage returns a subset of the database entries chosen as follows:
-//
-// * if (perPage * page) <= databaseSize, the function returns the `i`-th
-// sublist of size `k` of the database entries, where `i = page - 1` and
-// `k = perPage`. E.g. if the database contains a set of entries with
-// descriptors [1, 2, 3, 4] and (page, perPage) = (2, 2), the returned list
-// will contain the elements with descriptors [3, 4].
+// ChannelPage returns the requested `page` of the database entries.
 //
-// * if (perPage * (page-1)) < databaseSize < (perPage * page), the function
-// returns the last (databaseSize - perPage * (page-1)) items. E.g. if the
-// database contains a set of entries with descriptors [1, 2, 3, 4] and
-// (page, perPage) = (2, 3), the returned list will contain the element with
-// descriptor 4.
-//
-// * otherwise, the function returns an empty list (due to the page index being
-// out-of-bounds).
+// Given the entire database as a list of channels, the elements starting from
+// index `page*perPage` are inserted in the list sequentially until either the
+// list contains `perPage` elements, or there are no more database entries.
 //
+// If the database has less than `page*perPage` elements,
+// an empty list is returned.
 //
 // ChannelPage ensures:
-// * len(channels) <= int(perPage)
+// * err != nil || len(channels) <= int(perPage)
+// * err != nil || !contracts.InTest || ( t.mustCount() > uint64((page-1) * perPage) || len(channels) == 0 )
+// * err != nil || !contracts.InTest || ( t.mustCount() <= uint64((page-1) * perPage) || len(channels) > 0 )
+// * err != nil || !contracts.InTest || ( t.mustCount() < uint64(page * perPage) || len(channels ) == int(perPage) )
```

Comment:
These contracts are a bit hard to read since -- what about involving a NOT to signal the implication, _e.g.,_:
```
* err != nil || !contracts.InTest || !( t.mustCount() <= uint64((page-1) * perPage) || len(channels) == 0
```

I'd rename the package "contracts" into "dbc" to save a couple of characters.